### PR TITLE
Auto-capacity-allocation

### PR DESCRIPTION
**Pull request by petrkotas**
### Which issue this PR addresses:

Fixes https://dev.azure.com/msazure/AzureRedHatOpenShift/_workitems/edit/9673960/

### What this PR does / why we need it:

Introduce new feature flag to enable AutoSizedNodes. This enables the openshift feature to compute
resource allocation dynamically based on the node size. It does that by creating Kubeletconfig config with the feature turned on.

It also removes the aro-limits KubeletConfig that will be conflicting.

### CODE REVIEW

File: pkg/operator/controllers/configs/config_autonodesize.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		// the right config is there delete
+		err = r.Delete(ctx, &config, &client.DeleteOptions{})
+	}
+
+	// return error only if it is not NotFound
+	return client.IgnoreNotFound(err)
+}
+
+func (c *autoNodeSizeConfig) AddOwns(builder *builder.Builder) *builder.Builder {
+	return builder.Owns(&mcv1.KubeletConfig{})
```
Comment (m1kola) [inline pkg/operator/controllers/configs/config_autonodesize.go:86]: Should we add a predicate to make sure that we only reconcile `configName` and don't care about rest of the `KubeletConfig` objects?

This also removes the need to do this all the time:

```go
key := types.NamespacedName{
	Name: configName,
}
```

We shoud be able to just use `ctrl.Request` which we receive in `Reconcile`.

File: pkg/operator/controllers/configs/config_autonodesize.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+func (c *autoNodeSizeConfig) IsApplicable(aro arov1alpha1.Cluster, r *Reconciler, ctx context.Context) bool {
+	var config mcv1.KubeletConfig
+	var err error
+
+	key := types.NamespacedName{
+		Name: configName,
+	}
+
+	if aro.Spec.Features.ReconcileAutoSizedNodes {
```
Comment (m1kola) [inline pkg/operator/controllers/configs/config_autonodesize.go:48]: It is `false` by default. Do we want to update `pkg/operator/deploy/deploy.go` to enable it?

File: pkg/operator/controllers/configs/doc.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the Apache License 2.0.
+
+// Configs controller allows to create custom config that would be applied
+// into the cluster.
+// The design is inspired by the workaround controller. The difference is
+// the focus on configuration instead of hotfixes.
+// It also allows for flexible applicable criteria checks since it is
+// not fixed on any single item. All checks are done in the Config itself.
+// The configs operator takes the cluster config as a main object.
+// Other related objects are added in individual configs.
```
Comment (m1kola) [inline pkg/operator/controllers/configs/doc.go:13]: I'm not convinced that we should implement a generic config controller at this point.

I would probably start with a controller which does one thing: reconciles one specific kubeletconfig. If/when we see more use cases for this we can reconsider having a generic controller for configs.

**Review by m1kola — CHANGES_REQUESTED** [review]
Question: when we create/delete these `KubeletConfig` with `AutoSizingReserved`  - what happens? Will it rotate the node/restart kubelet?

**Comment by petrkotas** [comment]
I experienced multiple issues when reworking the PR, closing in favor of https://github.com/Azure/ARO-RP/pull/1862 which reworks the item mentions in this PR.

**Pull request closed without merging**